File: projects/snake/snake_mcp_server.py
"""
贪吃蛇 FastMCP 服务器
通过 IPC 文件与运行中的游戏通信，提供工具供 AI 调用

用法:
  python snake_mcp_server.py

在 VS Code 的 mcp.json 中注册:
  "snake-game": {
    "command": "E:\\DemoStudio\\.venv\\Scripts\\python.exe",
    "args": ["projects/snake/snake_mcp_server.py"],
    "cwd": "E:\\DemoStudio"
  }
"""
import sys
import os
from pathlib import Path

# ensure root and projects/snake/ are importable
_root = Path(__file__).resolve().parent.parent.parent
sys.path.insert(0, str(_root))
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from fastmcp import FastMCP
import snake_shared
from core.logger import get_logger

logger = get_logger('mcp')

# 创建 MCP 服务器实例
mcp = FastMCP(
    "贪吃蛇游戏控制器",
)


@mcp.tool(name="get_state", description="获取游戏当前状态：分数、蛇位置、食物位置、游戏是否结束等")
def get_state() -> dict:
    """获取游戏当前完整状态"""
    state = snake_shared.read_state()
    if state is None:
        return {
            "status": "error",
            "message": "游戏未运行。请先双击 snake_game.bat 启动游戏窗口。"
        }
    # 移除 body 大数组避免消息过长，单独提供摘要
    body = state.get("body", [])
    head = state.get("head", "?")
    length = state.get("snake_length", 0)
    state_display = {
        "status": "running",
        "score": state.get("score", 0),
        "snake_length": length,
        "head_position": head,
        "body_preview": body[:5],
        "body_count": len(body),
        "food_position": state.get("food"),
        "direction": state.get("direction"),
        "game_over": state.get("game_over", False),
        "grid_size": state.get("grid_size", 20),
    }
    return state_display


@mcp.tool(name="get_body", description="获取蛇身的完整坐标列表")
def get_body() -> dict:
    """获取蛇身的完整坐标"""
    state = snake_shared.read_state()
    if state is None:
        return {"status": "error", "message": "游戏未运行"}
    return {
        "status": "running",
        "body": state.get("body", []),
        "head": state.get("head"),
        "snake_length": state.get("snake_length", 0),
    }


@mcp.tool(name="get_grid", description="获取整个网格的状态 (0=空地, 1=蛇身, 2=蛇头, 3=食物)")
def get_grid() -> dict:
    """获取网格地图"""
    state = snake_shared.read_state()
    if state is None:
        return {"status": "error", "message": "游戏未运行"}

    grid_size = state.get("grid_size", 20)
    half = grid_size // 2

    # 初始化网格
    grid = [[0] * grid_size for _ in range(grid_size)]

    # 标记蛇身
    body = state.get("body", [])
    for x, z in body:
        gx = x + half
        gz = z + half
        if 0 <= gx < grid_size and 0 <= gz < grid_size:
            grid[gz][gx] = 1

    # 标记蛇头
    head = state.get("head")
    if head:
        hx, hz = head
        ghx = hx + half
        ghz = hz + half
        if 0 <= ghx < grid_size and 0 <= ghz < grid_size:
            grid[ghz][ghx] = 2

    # 标记食物
    food = state.get("food")
    if food:
        fx, fz = food
        gfx = fx + half
        gfz = fz + half
        if 0 <= gfx < grid_size and 0 <= gfz < grid_size:
            grid[gfz][gfx] = 3

    return {
        "status": "running",
        "grid": grid,
        "grid_size": grid_size,
        "legend": {"0": "空地", "1": "蛇身", "2": "蛇头", "3": "食物"},
    }


@mcp.tool(name="set_direction", description="设置蛇的移动方向 (up/down/left/right)")
def set_direction(direction: str) -> dict:
    """切换蛇的前进方向"""
    valid = ["up", "down", "left", "right"]
    if direction not in valid:
        return {
            "status": "error",
            "message": f"无效方向 '{direction}'，可用: {', '.join(valid)}"
        }
    snake_shared.write_command("set_direction", direction=direction)
    return {"status": "ok", "direction": direction, "message": f"方向已切换为 {direction}"}


@mcp.tool(name="restart", description="重新开始游戏")
def restart() -> dict:
    """重启游戏"""
    snake_shared.write_command("restart")
    return {"status": "ok", "message": "游戏已重新开始"}


@mcp.tool(name="set_speed", description="设置游戏速度 (0.05~0.50, 越小越快)")
def set_speed(speed: float) -> dict:
    """调整游戏移动间隔"""
    speed = max(0.05, min(0.50, speed))
    snake_shared.write_command("set_speed", speed=speed)
    speed_label = "极快" if speed <= 0.08 else ("快" if speed <= 0.12 else (
        "中" if speed <= 0.2 else "慢"))
    return {
        "status": "ok",
        "speed": speed,
        "label": speed_label,
        "message": f"速度已调整为 {speed_label} ({speed:.2f})",
    }


@mcp.tool(name="get_log", description="获取最近的游戏事件日志")
def get_log(limit: int = 20) -> dict:
    """读取游戏日志"""
    lines = snake_shared.read_log(limit=limit)
    return {
        "status": "ok",
        "count": len(lines),
        "logs": lines,
        "log_file": str(snake_shared.LOG_FILE),
    }


@mcp.tool(name="help", description="显示所有可用工具及其说明")
def help() -> str:
    """返回帮助信息"""
    return """贪吃蛇游戏控制器 - 可用工具:

1. get_state        - 获取游戏状态 (分数、蛇位置、食物等)
2. get_body         - 获取蛇身完整坐标列表
3. get_grid         - 获取 20x20 网格地图
4. set_direction    - 设置蛇方向 (up/down/left/right)
5. restart          - 重新开始游戏
6. set_speed        - 调整速度 (0.05~0.50, 越快值越小)
7. get_log          - 获取最近游戏事件日志
8. help             - 显示此帮助

使用示例: 调用 set_direction(direction="right") 让蛇向右转
"""


if __name__ == "__main__":
    logger.info("=" * 50)
    logger.info("Snake MCP Server started")
    logger.info("IPC 目录: {}", snake_shared.IPC_DIR)
    logger.info("=" * 50)
    logger.info("State file: snake_state.json")
    logger.info("Command file: snake_ipc.json")
    logger.info("Log file: snake_log.txt")
    logger.info("=" * 50)
    logger.info("Tip: run snake_game.bat first to start the game window")
    logger.info("=" * 50)
    mcp.run(transport="stdio")

File: projects/snake/snake-mcp-server.mjs
/**
 * Snake Game MCP 服务器（基于 @modelcontextprotocol/sdk）
 * 独立于编辑器 MCP，提供贪吃蛇游戏专用控制工具
 *
 * VS Code MCP 配置 (.vscode/mcp.json):
 *   "snake-game": {
 *     "type": "stdio",
 *     "command": "node",
 *     "args": ["projects/snake/snake-mcp-server.mjs"],
 *     "cwd": "E:\\DemoStudio"
 *   }
 */
import { Server } from '@modelcontextprotocol/sdk/server/index.js'
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js'
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js'

const EDITOR_API = 'http://127.0.0.1:9877'

async function callEditor(command, params = {}) {
  try {
    const resp = await fetch(`${EDITOR_API}/api/command`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ command, params }),
    })
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`)
    return await resp.json()
  } catch (err) {
    return { status: 'error', message: `编辑器不可达: ${err.message}` }
  }
}

async function getGameState() {
  try {
    const resp = await fetch(`${EDITOR_API}/api/game-state`)
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`)
    return await resp.json()
  } catch (err) {
    return { status: 'error', message: `编辑器不可达: ${err.message}` }
  }
}

async function getStatus() {
  try {
    const resp = await fetch(`${EDITOR_API}/api/status`)
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`)
    return await resp.json()
  } catch (err) {
    return { status: 'error', message: `编辑器不可达: ${err.message}` }
  }
}

// ─── 创建 MCP 服务器 ───

const server = new Server(
  { name: 'snake-game', version: '1.0.0' },
  { capabilities: { tools: {} } },
)

// ─── 列出工具 ───

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [
    {
      name: 'start_game',
      description: '启动贪吃蛇游戏',
      inputSchema: { type: 'object', properties: {} },
    },
    {
      name: 'stop_game',
      description: '停止贪吃蛇游戏',
      inputSchema: { type: 'object', properties: {} },
    },
    {
      name: 'toggle_game',
      description: '切换游戏启动/停止',
      inputSchema: { type: 'object', properties: {} },
    },
    {
      name: 'send_input',
      description: '发送方向键控制蛇移动',
      inputSchema: {
        type: 'object',
        properties: {
          key: {
            type: 'string',
            description: '方向键: ArrowUp / ArrowDown / ArrowLeft / ArrowRight',
          },
        },
        required: ['key'],
      },
    },
    {
      name: 'get_game_state',
      description: '获取贪吃蛇实时状态（蛇头位置、分数、方向、食物位置等）',
      inputSchema: { type: 'object', properties: {} },
    },
  ],
}))

// ─── 调用工具 ───

server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params

  switch (name) {
    case 'start_game': {
      const result = await callEditor('start_game')
      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }],
      }
    }
    case 'stop_game': {
      const result = await callEditor('stop_game')
      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }],
      }
    }
    case 'toggle_game': {
      const status = await getStatus()
      if (status.gameRunning) {
        const result = await callEditor('stop_game')
        return {
          content: [{ type: 'text', text: JSON.stringify(result, null, 2) }],
        }
      } else {
        const result = await callEditor('start_game')
        return {
          content: [{ type: 'text', text: JSON.stringify(result, null, 2) }],
        }
      }
    }
    case 'send_input': {
      const key = args?.key || ''
      const result = await callEditor('send_input', { key })
      return {
        content: [{ type: 'text', text: JSON.stringify(result, null, 2) }],
      }
    }
    case 'get_game_state': {
      const result = await getGameState()
      const data = result.data
      if (!data) {
        return {
          content: [{ type: 'text', text: JSON.stringify({
            status: 'ok',
            message: '游戏未运行，没有实时数据',
            data: null,
          }, null, 2) }],
        }
      }
      // 格式化输出为更可读的结构
      const formatted = {
        status: 'ok',
        game: {
          phase: data.phase,
          score: data.score,
          snakeLength: data.snakeLength,
          headPosition: data.headPosition,
          foodPosition: data.foodPosition,
          currentDirection: data.currentDirection,
        },
        summary: data.phase === 'gameover'
          ? `💀 Game Over · 得分: ${data.score} · 蛇长: ${data.snakeLength}`
          : data.phase === 'playing'
            ? `🐍 运行中 · 得分: ${data.score} · 蛇长: ${data.snakeLength} · 蛇头: (${data.headPosition?.x}, ${data.headPosition?.z}) · 食物: (${data.foodPosition.x}, ${data.foodPosition.z})`
            : `⏳ 等待中`,
      }
      return {
        content: [{ type: 'text', text: JSON.stringify(formatted, null, 2) }],
      }
    }
    default:
      throw new Error(`未知工具: ${name}`)
  }
})

// ─── 启动 STDIO 传输 ───

async function main() {
  const transport = new StdioServerTransport()
  await server.connect(transport)
  console.error('[MCP] Snake Game MCP Server 已启动')
}

main().catch((err) => {
  console.error('[MCP] 启动失败:', err)
  process.exit(1)
})
